syntax = "proto3";

service BalanceService {
    rpc UpdateBalance(UpdateBalanceRequest) returns (BalanceResponse);
    rpc OpenTransaction(OpenTransactionRequest) returns (TransactionResponse);
    rpc CommitTransaction(TransactionRequest) returns (TransactionStatus);
    rpc CancelTransaction(TransactionRequest) returns (TransactionStatus);
}

message UpdateBalanceRequest {
    int32 user_id = 1;
    float current_change = 2;
    float max_change = 3;
}

message BalanceResponse {
    float current_balance = 1;
    float max_balance = 2;
}

message OpenTransactionRequest {
    int32 user_id = 1;
    float amount = 2;
}

message TransactionResponse {
    int32 transaction_id = 1;
    float locked_amount = 2;
}

message TransactionRequest {
    int32 transaction_id = 1;
}

message TransactionStatus {
    string status = 1;
}